rust/private/clippy.bzl

```diff
@@ -0,0 +1,167 @@
+    crate_type = getattr(ctx.rule.attr, "crate_type") if hasattr(ctx.rule.attr, "crate_type") else "lib"
+    crate_info = CrateInfo(
+        name = crate_name,
+        type = crate_type,
+        root = crate_root_src(ctx, srcs = rust_srcs),
+        srcs = rust_srcs,
+        deps = ctx.rule.attr.deps,
+        proc_macro_deps = ctx.rule.attr.proc_macro_deps,
+        aliases = ctx.rule.attr.aliases,
+        output = None,
+        edition = get_edition(ctx.rule.attr, toolchain),
+        rustc_env = ctx.rule.attr.rustc_env,
+    )
+
+    dep_info, build_info = collect_deps(
+        ctx.label,
+        crate_info.deps,
+        crate_info.proc_macro_deps,
+        crate_info.aliases,
+        toolchain,
+    )
+
+    compile_inputs, out_dir = collect_inputs(
+        ctx,
+        toolchain,
+        crate_info,
+        dep_info,
+        build_info
+    )
+
+    args, env = construct_arguments(
+        ctx,
+        toolchain,
+        crate_info,
+        dep_info,
+        output_hash = None,
+        rust_flags = [])
+
+    # A marker file indicating clang tidy has executed successfully.
+    # This file is necessary because "ctx.actions.run" mandates an output.
+    clippy_marker = ctx.actions.declare_file(ctx.label.name + "_clippy.ok")
+
+    command = construct_compile_command(
+        ctx,
+        toolchain.clippy_driver.path,
+        toolchain,
+        crate_info,
+        build_info,
+        out_dir,
+    ) + (" && touch %s" % clippy_marker.path)
+
+    # Deny the default-on clippy warning levels.
+    #
+    # If these are left as warnings, then Bazel will consider the execution
+    # result of the aspect to be "success", and Clippy won't be re-triggered
+    # unless the source file is modified.
+    args.add("-Dclippy::style")
+    args.add("-Dclippy::correctness");
+    args.add("-Dclippy::complexity");
+    args.add("-Dclippy::perf");
+
+    ctx.actions.run_shell(
+        command = command,
+        inputs = compile_inputs,
+        outputs = [clippy_marker],
+        env = env,
+        tools = [toolchain.clippy_driver],
+        arguments = [args],
+        mnemonic = "Clippy",
+    )
+
+    return [
+        OutputGroupInfo(clippy_checks = depset([clippy_marker])),
+    ]
+
+# Executes a "compile command" on all specified targets.
+#
+# Example: Run the clang-tidy checker on all targets in the codebase.
+#   bazel build --aspects=@rules_rust//rust:clippy.bzl%rust_clippy_aspect \
+#               --output_groups=clippy_checks \
+#               //...
+rust_clippy_aspect = aspect(
+    attr_aspects = ["deps"],
+    fragments = ["cpp"],
+    host_fragments = ["cpp"],
+    attrs = {
+        "_cc_toolchain": attr.label(
+            default = Label("@bazel_tools//tools/cpp:current_cc_toolchain"),
+        ),
+    },
+    toolchains = [
+        "@io_bazel_rules_rust//rust:toolchain",
+        "@bazel_tools//tools/cpp:toolchain_type"
+    ],
+    implementation = _clippy_aspect_impl,
+)
```

Comment:
Just a though: 
I don't remember the syntax exact of aspect but I remember they could be propagated along all the edges. If this is the case you are going to flatten all the sources depset of all the targets.

If you understand more how that works, can you expand? Else I'll try to go back to the doc to investigate.

Reply:
(warning: this is one of the first aspects I've written, so I am far from an expert, and would appreciate all the guidance I can get)

So I'm aware that aspects *can* propagate along all edges, but I'm not actually certain that is desired behavior here.

In the documentation https://docs.bazel.build/versions/master/skylark/aspects.html , I know that a somewhat recursive notation allows this propagation:

```
def _file_count_aspect_impl(target, ctx):
    ... (redacted)
    # Get the counts from our dependencies.
    for dep in ctx.rule.attr.deps:
        count = count + dep[FileCountInfo].count
    return [FileCountInfo(count = count)]

file_count_aspect = aspect(
    implementation = _file_count_aspect_impl,
    attr_aspects = ['deps'],
    ... (redacted)
 )
```

In this example, accessing the `dep[FileCountInfo]` value accesses the propagated return value of the aspect applied to a dependency.

In the case of the clippy aspect, however, I don't think we actually want to run clippy on all dependencies. For most crates, this would hop into third party dependencies, where clippy would possibly fail, but without really providing actionable feedback to the caller.

Reply:
I think you need to remove the `attr_aspects` attribute to the aspect. The rule declare in which way the aspect propagate along the top edge, so no need for the aspect to propagate further.

Indeed for clippy that would be strange to evaluate the source of all recursive targets.

Reply:
Done